v0.26.0 - Major Release (UX and polish focused)

TkTech opened the pull request:
(This is a huge release, changelog to come. WIP)

0.26.0
------

✨ Improvements

- Reworked the dashboard, much more informative for operational insights, light/dark themes.
- Added much, much better composable filtering widgets to both jobs and workflows.
- Added a visualization of your upcoming (and past) cron schedule, making it easy to identify overloaded periods.
- Added a realtime view of the internal Chancy communication traffic, helping to debug and identify pain points (like using `push()` in a loop instead of a single `push_many()`)
- HTTP API now supports mutation - create queues on the fly, move jobs around, retry workflows, etc...
- HTTP API is now stable, with documentation still a work in progress.
- HTTP API now includes an unauthenticated endpoint for health checks, meant for docker/containers (#54)
- Added dozens of new metrics, including metrics for the metrics plugin and metrics on workflows.
- Added postgres 18 to test matrix (@jklaise)
- The starlette app can now be created without being started for advanced use cases, and the mounting prefix for the SPA can be changed (@jklaise, #64)
- Workflow performance has been significantly improved and can now handle many, individual workflows with thousands of jobs without issue.
- Workflows will now be validated before saving, verifying there are no unreachable nodes in the DAG, no cycles, etc
- Can now be used as a django-tasks backend, albeit with less functionality than using Chancy directly

🐛 Fixes

- Respect CLI flags for API plugin configuration by @alfawal (#68)

<img width="2534" height="1345" alt="image" src="https://github.com/user-attachments/assets/021d4f3a-269c-4bef-ab43-a6f2e8bbfcd7" />
<img width="2534" height="1345" alt="image" src="https://github.com/user-attachments/assets/56ddb69e-4cae-4f99-9b81-f6829b903da2" />
<img width="2534" height="1345" alt="image" src="https://github.com/user-attachments/assets/7c4d4d7f-7e62-4751-bf05-c071fa943e57" />
<img width="2534" height="1345" alt="image" src="https://github.com/user-attachments/assets/a4b6e575-73ca-4ac2-ad22-1367ae28e336" />
<img width="2542" height="1251" alt="image" src="https://github.com/user-attachments/assets/83cfe54c-8cae-4560-bbae-e4b57c18dc4b" />

TkTech committed:
Cleanups

TkTech committed:
useJob should periodically refresh if the job isn't finished.

TkTech committed:
Add workflow metric visualization.

TkTech committed:
Switch to umami for doc site analytics.

TkTech commented:
Added a cron job visualizer.

<img width="2544" height="1242" alt="image" src="https://github.com/user-attachments/assets/38ebd3b3-6b36-465a-ac8b-652eaaedb6de" />

alfawal committed:
Respect CLI flags for API plugin configuration

API plugin settings (host, port, debug, allow_origin) are now overridden if the corresponding CLI flags are explicitly provided, even when an Api instance is defined outside the default behaviour.

fcortes commented:
Hi! I saw that you mentioned django-tasks compatibility on this PR but couldn't find any reference about it on the code. Do you mind committing those changes or maybe open a new PR with it (to avoid getting this one so large). I was playing around with chancy and also have a semi-functional implementation and it would be nice to see how you managed some issues I've been encountering.

Great project btw!

TkTech committed:
Merge pull request #68 from alfawal/respect-explicit-cli-flags

Respect CLI flags for API plugin configuration

phillycheeze commented:
@TkTech We just started leveraging chancy and super excited for this new dashboard! Any outstanding items you're looking for help with?

TkTech commented:
> @TkTech We just started leveraging chancy and super excited for this new dashboard! Any outstanding items you're looking for help with?

It's been running for testing in a production environment since Monday, so I'd say we'll release this weekend if all goes well.

TkTech commented:
> Hi! I saw that you mentioned django-tasks compatibility on this PR but couldn't find any reference about it on the code. Do you mind committing those changes or maybe open a new PR with it (to avoid getting this one so large). I was playing around with chancy and also have a semi-functional implementation and it would be nice to see how you managed some issues I've been encountering.
> 
> Great project btw!

I've left some comments in the #82 ticket regarding this, there are some behavioral Qs to answer before we commit to anything.

TkTech committed:
Hacky attempt at django-tasks support.

TkTech committed:
Merge remote-tracking branch 'origin/26_ux' into 26_ux

TkTech committed:
UUID-Everywhere

We had partially transitioned to UUIDs in v0.25, this completes the work and adds a regression test and closes #84. Also adds some Executor utilities to the public API for getting and iterating the current job pool.

TkTech committed:
Prepare for an upcoming 3.16 deprecation in the asyncio module.

TkTech committed:
The Mac OS CI/CD runners are very unreliable, 5x the very tight timing test in the test_scale regression catcher.

TkTech committed:
Cover async executor in the regression test as well, since it also supports cancellation.

TkTech committed:
Collapse regression test into old, flawed test. Try to improve performance of two slow tests.

TkTech committed:
Parallelize the test runners, and fix a constraint missing a table prefix we found along the way.

TkTech committed:
Skip process cancellation on Windows (no signals)

TkTech committed:
Merge pull request #85 from TkTech/finish_uuid_transition

UUID-Everywhere
